application/Common/HookLibrary/dllmain.cpp

```diff
@@ -0,0 +1,193 @@
+// dllmain.cpp : Defines the entry point for the DLL application.
+#include "pch.h"
+#include "dllmain.h"
+
+BOOL APIENTRY DllMain(HANDLE hModule, DWORD  ul_reason_for_call,
+    LPVOID lpReserved)
+{
+    switch (ul_reason_for_call)
+    {
+    case DLL_PROCESS_ATTACH:
+        
+        hInstHookDll = (HINSTANCE)hModule;
+        break;
+    }
+    return TRUE;
+}
+
+LRESULT CALLBACK GetMsgProc(int nCode, WPARAM wParam, LPARAM lParam)
+//this is the hook procedure
+{
+    LRESULT retVal = CallNextHookEx(GetMessageHook, nCode, wParam, lParam);
+    //a pointer to hold the MSG structure that is passed as lParam
+    MSG* msg;
+    if (semaphore == NULL) {
+        semaphore = CreateSemaphore(NULL, 0, BUFFERSIZE, TEXT(SEMAPHORE_GUID));
+    }
+
+    //lParam contains pointer to MSG structure.
+    msg = (MSG*)lParam;
+    /* see https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644981(v=vs.85) */
+    if (nCode >= 0 && nCode == HC_ACTION && msg->message < MESSAGETABLESIZE && messageHasListener[msg->message])
+    {
+        unsigned int bufferSlot = InterlockedIncrement(&globalBufferIterator) % BUFFERSIZE; //atomic increment. overflows should not matter as we modulo it anyways.
+        unsigned int type = msg->message;
+        globalTimeStamps[bufferSlot] = msg->time;
+        if ((type >= WM_MOUSEMOVE && type <= WM_MOUSEHWHEEL) || type == WM_KEYDOWN) {
+            globalMessageBuffer[bufferSlot] = { msg->message, msg->hwnd, msg->wParam, {0} };
+            globalMessageBuffer[bufferSlot].data[0] = msg->pt.x;
+            globalMessageBuffer[bufferSlot].data[1] = msg->pt.y;
+        }
+        else
+        {
+            goto forwardEvent;
+        }
+        ReleaseSemaphore(semaphore, 1, NULL); /* mark one message as available to the dispatcher running in the MORR AS */
+    }
+forwardEvent:
+    /* pass message onto target application */
+    return retVal;
```

Comment:
As i might add additional `else if` here, i didn't want to move it inside as then i'd have to duplicate it everywhere later. But i should change the error handling here (else should never occur), because currently this still increases the global iterator even if `else` is hit.